cpp/ql/src/semmle/code/cpp/models/interfaces/PointerWrapper.qll

```diff
@@ -0,0 +1,12 @@
+/** Provides classes for modeling pointer wrapper types and expressions. */
+
+private import cpp
+
+/** A class that wraps a pointer type. For example, `std::unique_ptr` and `std::shared_ptr`. */
+abstract class PointerWrapper extends Class {
+  /** Gets a member fucntion of this class that dereferences wrapped pointer, if any. */
+  abstract MemberFunction getADereferenceFunction();
+
+  /** Gets a member fucntion of this class that returns the wrapped pointer, if any. */
+  abstract MemberFunction getAnUnwrapperFunction();
```

Comment:
These two members look very similar. Is the difference just about whether they return a pointer or a reference? If so, I'd expect their names to reflect that. They could also be just one function, and the caller would distinguish based on types where it mattered (it shouldn't matter for IR, right?).

Reply:
That's true. My intention was that `getADereferenceFunction` modeled an overloaded `operator*` that dereferenced the underlying pointer (i.e., `*my_smart_pointer`). But in reality, I guess these functions always return references to the wrapped pointer, and the dereference is really done by a `ReferenceDereferenceExpr` (or a `LoadInstruction`) on the caller side.

Reply:
Fixed in b466f0515d182cde679a9beacd5a88ed6b5c1376.

I didn't add the extra condition to the `OverloadedPointerDereferenceFunction` disjunction of `getAnUnwrapperFunction` that checks that the return type is actually a reference since this _is_ the case for the specific smart pointers we plan to model in this iteration. But it's something that we can revisit later.